framework/decode/vulkan_swapchain.cpp

```diff
@@ -0,0 +1,168 @@
+**
+** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+** IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+** FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+** AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+** LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+** FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+** DEALINGS IN THE SOFTWARE.
+*/
+
+#include "decode/vulkan_swapchain.h"
+
+GFXRECON_BEGIN_NAMESPACE(gfxrecon)
+GFXRECON_BEGIN_NAMESPACE(decode)
+
+const int32_t  kDefaultWindowPositionX = 0;
+const int32_t  kDefaultWindowPositionY = 0;
+const uint32_t kDefaultWindowWidth     = 320;
+const uint32_t kDefaultWindowHeight    = 240;
+
+void VulkanSwapchain::Clean()
+{
+    if (options_surface_index_ >= create_surface_count_)
+    {
+        GFXRECON_LOG_WARNING("Rendering was restricted to surface index %u, but a surface was never created for that "
+                             "index; replay created %d surface(s)",
+                             options_surface_index_,
+                             create_surface_count_);
+    }
+
+    // Destroy any windows that were created for Vulkan surfaces.
+    for (auto window : active_windows_)
+    {
+        auto wsi_context    = application_ ? application_->GetWsiContext(window->GetWsiExtension()) : nullptr;
+        auto window_factory = wsi_context ? wsi_context->GetWindowFactory() : nullptr;
+        GFXRECON_ASSERT(window_factory);
+        window_factory->Destroy(window);
+    }
+}
+
+VkResult VulkanSwapchain::CreateSurface(InstanceInfo*                       instance_info,
+                                        const std::string&                  wsi_extension,
+                                        VkFlags                             flags,
+                                        HandlePointerDecoder<VkSurfaceKHR>* surface,
+                                        const encode::InstanceTable*        instance_table,
+                                        application::Application*           application,
+                                        int32_t                             options_surface_index)
+{
+    assert(instance_info != nullptr);
+
+    instance_table_        = instance_table;
+    application_           = application;
+    options_surface_index_ = options_surface_index;
+
+    VkInstance    instance       = instance_info->handle;
+    VkSurfaceKHR* replay_surface = nullptr;
+    VkResult      result         = VK_SUCCESS;
+
+    if (surface != nullptr)
+    {
+        replay_surface = surface->GetHandlePointer();
+    }
+
+    // For multi-surface captures, when replay is restricted to a specific surface, only create a surface for
+    // the specified index.
+    if ((options_surface_index_ == -1) || (options_surface_index_ == create_surface_count_))
+    {
+        // Create a window for our surface.
```

Comment:
None of this is really necessary for Offscreen, right?  If so, why don't we add that scenario to the if check above?

Reply:
`VulkanOffscreenSwapchain::CreateSurface` overrides it. Is that your meaning?